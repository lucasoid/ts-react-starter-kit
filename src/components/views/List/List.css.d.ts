export const listItem: string;
export const category: string;
export const name: string;
export const quantity: string;
export const active: string;
export const inactive: string;
