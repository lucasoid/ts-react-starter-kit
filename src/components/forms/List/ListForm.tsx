import * as React from 'react';
import uuidv4 from 'uuid/v4';
import { Form } from '~components/forms/Form';
import { TextField } from '~components/forms/Inputs';

interface IListFormProps {
    onSubmit: Function;
    onCancel: Function;
}
export const ListForm: React.FC<IListFormProps> = props => {
    const [name, setName] = React.useState('');
    const [members, setMembers] = React.useState('');

    return (
        <Form
            onSubmit={() => {
                props.onSubmit({
                    id: uuidv4(),
                    name: name,
                    owner: 'me@domain.com',
                    members: members.split(','),
                });
            }}
            onCancel={props.onCancel}
            submitText={'Create'}
            cancelText={'Cancel'}
        >
            <TextField value={name} label="List name" onChange={setName} />
            <TextField value={'me@domain.com'} label="List owner" onChange={setName} disabled={true} />
            <TextField value={members} label="Share with" onChange={setMembers} />
        </Form>
    );
};
