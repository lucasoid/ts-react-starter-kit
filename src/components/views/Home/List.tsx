import * as React from 'react';
import { Link } from 'react-router-dom';
import { Loading } from '~components/loading';
import * as styles from './Home.css';
import { ThemeContext, Themes } from '~theme';
import { IList } from '~services/shoppingListApi';

interface IListProps {
    list: IList;
}
export const List: React.FC<IListProps> = props => {
    let { theme } = React.useContext(ThemeContext);

    return (
        <div className={`${styles.menuItem} ${theme === Themes.DARK ? styles.dark : styles.light}`}>
            {!props.list ? (
                <>
                    <Loading el={'a'} lines={1} width={'35%'} />
                    <Loading el={'aside'} lines={1} width={'25%'} />
                </>
            ) : (
                <>
                    <Link to={`/lists/${props.list.id}`} title={props.list.name}>
                        {props.list.name}
                    </Link>
                    <aside>Created by {props.list.owner}</aside>
                </>
            )}
        </div>
    );
};
