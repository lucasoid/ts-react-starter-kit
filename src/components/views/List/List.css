.listItem {
    padding: 0.5em;
    display: flex;
    flex-direction: row;
}

.listItem.inactive {
    text-decoration: line-through;
}

.listItem.active {
    text-decoration: none;
}

.listItem .category {
    flex: 0 0 30%;
}

.listItem .name {
    flex: 0 0 30%;
}

.listItem .quantity {
    flex: 0 0 10%;
}
