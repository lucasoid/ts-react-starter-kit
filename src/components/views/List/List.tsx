import * as React from 'react';
import { Link, /*useParams,*/ withRouter } from 'react-router-dom';
import { View } from '~components/views/View';
import { Button } from '~components/ui/Button';
import {
    fetchListItems,
    IListItem,
    subscribeToListItems,
    unsubscribeToListItems,
    IList,
    fetchLists,
} from '~services/shoppingListApi';
import * as styles from './List.css';

interface IListProps {
    match?: any;
}

interface IListState {
    list: IList;
    items: IListItem[];
}

class InertList extends React.Component<IListProps, IListState> {
    state = {
        list: {
            id: null,
            name: null,
            owner: null,
            members: [],
        },
        items: [],
    };

    subscriber = items => {
        this.setState({ items: items });
    };

    async componentDidMount() {
        const { listId } = this.props.match.params;
        let lists = await fetchLists();
        let list = lists.find(_list => _list.id === listId);
        this.setState({ list }, () => {
            subscribeToListItems(this.state.list.id, this.subscriber);
            fetchListItems(this.state.list.id);
        });
    }

    componentWillUnmount() {
        unsubscribeToListItems(this.state.list.id, this.subscriber);
    }

    render() {
        return (
            <View>
                <Link to="/">‚Üê All lists</Link>
                <h2>{this.state.list.name}</h2>
                <aside>Owner: {this.state.list.owner}</aside>
                <aside>Members: {this.state.list.members.join(', ')}</aside>
                {this.state.items.map((item, index) => (
                    <div
                        className={`${styles.listItem} ${item.isActive ? styles.active : styles.inactive}`}
                        key={item.id + index}
                    >
                        <div className={styles.category}>{item.category}</div>
                        <div className={styles.name}>{item.name}</div>
                        <div className={styles.quantity}>{item.quantity}</div>
                    </div>
                ))}
                <Button styles={{ marginTop: '1em', backgroundColor: 'teal' }}>+ New item</Button>
            </View>
        );
    }
}

export const List = withRouter(InertList);
