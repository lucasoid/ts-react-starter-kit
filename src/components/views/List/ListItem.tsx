import * as React from 'react';
import { Button } from '~components/ui/Button';
import { Loading } from '~components/Loading';
import * as styles from './List.css';
import { ThemeContext, Themes } from '~theme';
import { IListItem } from '~services/shoppingListApi';

interface IListItemProps {
    item: IListItem;
    onClick: Function;
}
export const ListItem: React.FC<IListItemProps> = props => {
    let { item } = props;
    let { theme } = React.useContext(ThemeContext);

    return !item ? (
        <div className={`${styles.listItem} ${styles.active} ${theme === Themes.DARK ? styles.dark : styles.light}`}>
            <div className={styles.checkbox}>
                <Button plaintext={true} styles={{ outline: 'none' }}>
                    ☐
                </Button>
            </div>
            <div className={styles.name}>
                <Loading width={'100%'} lines={1} el={'span'} />
            </div>
            <div className={styles.category}>
                <Loading width={'100%'} lines={1} el={'span'} />
            </div>
            <div className={styles.quantity}>
                <Loading width={'100%'} lines={1} el={'span'} />
            </div>
        </div>
    ) : (
        <div
            className={`
                ${styles.listItem}
                ${item.isActive ? styles.active : styles.inactive}
                ${theme === Themes.DARK ? styles.dark : styles.light}
            `}
            onClick={() => props.onClick(item)}
        >
            <div className={styles.checkbox}>
                <Button plaintext={true} styles={{ outline: 'none' }}>
                    {item.isActive ? '☐' : '☑'}
                </Button>
            </div>
            <div className={styles.name}>{item.name}</div>
            <div className={styles.category}>{item.category}</div>
            <div className={styles.quantity}>{item.quantity}</div>
        </div>
    );
};
