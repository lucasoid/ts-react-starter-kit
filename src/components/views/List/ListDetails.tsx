import * as React from 'react';
import { Button, ButtonTypes } from '~components/ui/Button';
import { Loading, Await } from '~components/Loading';
import { ListItem } from './ListItem';
import { Modal, ModalWidth } from '~components/ui/Modal';
import { ListItemForm } from '~components/forms/ListItem';
import {
    fetchListItems,
    IListItem,
    subscribeToListItems,
    unsubscribeToListItems,
    IList,
    fetchLists,
    editListItem,
    createListItem,
} from '~services/shoppingListApi';
import { E404 } from '~components/errors/ErrorTypes';

interface IListDetailsProps {
    listId: string;
}

export const ListDetails: React.FC<IListDetailsProps> = props => {
    let [list, setList] = React.useState({ id: null, name: null, owner: null, members: [] } as IList);
    let [loading, setLoading] = React.useState(true);
    let [items, setItems] = React.useState([]);
    let [notFound, setNotFound] = React.useState(false);
    let [showCreateDialog, setShowCreateDialog] = React.useState(false);
    React.useEffect(() => {
        const subscriber = function(items) {
            setItems(items);
            setLoading(false);
        };
        fetchLists().then(lists => {
            let list = lists.find(_list => _list.id === props.listId);
            if (!list) {
                setNotFound(true);
            } else {
                setList(list);
                subscribeToListItems(props.listId, subscriber);
                fetchListItems(list.id);
            }
        });
        return () => unsubscribeToListItems(props.listId, subscriber);
    }, [props.listId]);

    function toggle(item: IListItem) {
        item.isActive = !item.isActive;
        // update the API, which will trigger the subscriber when it completes
        editListItem(item);
        // update state optimistically while we wait
        let _items = [...items];
        _items.splice(
            items.findIndex(_item => _item.id === item.id),
            1,
            item,
        );
        setItems(_items);
    }

    function handleCreate(item: IListItem) {
        item.listId = props.listId;
        createListItem(item);
        let _items = [...items];
        _items.push(item);
        setItems(_items);
        setShowCreateDialog(false);
    }

    if (notFound) throw new E404('List not found');

    return (
        <Await
            loaded={!loading}
            renderFallback={() => (
                <>
                    <Loading el={'h2'} width={'25%'} lines={1}></Loading>
                    <Loading el={'aside'} width={'25%'} lines={1}></Loading>
                    <Loading el={'aside'} width={'25%'} lines={1}></Loading>
                    <ListItem item={null} key={1} onClick={() => {}} />
                    <ListItem item={null} key={2} onClick={() => {}} />
                    <ListItem item={null} key={3} onClick={() => {}} />
                </>
            )}
        >
            <h2>{list.name}</h2>
            <aside>Owner: {list.owner}</aside>
            <aside>Members: {list.members.join(', ')}</aside>
            {items.map((item, index) => (
                <ListItem item={item} key={item.id} onClick={toggle} />
            ))}
            <Button onClick={() => setShowCreateDialog(true)} type={ButtonTypes.PRIMARY} styles={{ marginTop: '1em' }}>
                + New item
            </Button>
            {showCreateDialog && (
                <Modal width={ModalWidth.MEDIUM} onDismiss={() => setShowCreateDialog(false)}>
                    <ListItemForm onSubmit={handleCreate} onCancel={() => setShowCreateDialog(false)} />
                </Modal>
            )}
        </Await>
    );
};
