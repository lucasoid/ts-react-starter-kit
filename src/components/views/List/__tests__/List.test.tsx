import * as React from 'react';
import { shallow } from 'enzyme';
import { List } from '../List';
// imported for mocking
import { Link, useParams } from 'react-router-dom';

jest.mock('react-router-dom', () => ({
    Link: props => <a>{props.children}</a>,
    useParams: () => ({ listId: '1' }),
}));

describe('ListItem', () => {
    it('renders a view of a list page', () => {
        let wrapper = shallow(<List />);
        expect(wrapper).toMatchSnapshot();
    });
});
