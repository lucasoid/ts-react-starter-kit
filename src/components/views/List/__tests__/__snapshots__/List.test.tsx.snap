// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem renders a view of a list page 1`] = `
<View>
  <Link
    to="/"
  >
    ‚Üê All lists
  </Link>
  <ListDetails
    listId="1"
  />
</View>
`;
